onedrivegui: init at 1.0.3

chewblacka opened the pull request:
## Description of changes
onedrivegui init at 1.0.3
OneDriveGUI is a GUI for the linux OneDrive client.
https://github.com/bpozdena/OneDriveGUI

Closes #257520

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

chewblacka committed:
onedrivegui: init at 1.0.3

bertof commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,70 @@
+  nativeBuildInputs = [ copyDesktopItems ];
+
+  propagatedBuildInputs = [ onedrive ] ++
+  (with python3Packages; [
```
Use something like 

```nix
builtins.attrValues { inherit (python3Packages) pyside6 requests }
```

bertof commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,70 @@
+    })
+  ];
+
+    preBuild = ''
```
Requires formatting

bertof commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,70 @@
+  '';
+
+  postInstall = ''
+    mkdir $out/lib/OneDriveGUI && cp -r src/{resources,ui} $out/lib/OneDriveGUI/
```
Usually it's a better idea to only use link (ln) from the $src, this way you're not duplicating the files and you're leaving Nix the handling task.

bertof reviewed (changes requested):
Hi, the package seem mostly fine. You're missing an entry in the `all-packages` file, so that it can be imported in nixpkgs.
The package builds correctly and the application seem to run. I don't have a configuration ready to test it thoroughly.

chewblacka commented:
> Hi, the package seem mostly fine. You're missing an entry in the `all-packages` file, so that it can be imported in nixpkgs. The package builds correctly and the application seem to run. I don't have a configuration ready to test it thoroughly.

Many thanks for the detailed and helpful feedback! This is my first package init and is exactly the kind of valuable feedback I was hoping for. I've incorporated all your suggested changes. Also changed the `pname` to `onedrivegui` (instead of `OneDriveGUI`).

Do I still need to manually change `all-packages`? I wasn't sure what the current convention is for new packages. Looking through recently merged PRs they seem to be all over the shop. Some contain additions to `all-packages`, but most don't. If you could clarify it would be much appreciated!

Many thanks again.

bertof commented:
Hey, nice to see new contributors and contributions. I didn't knew of the per-name packaging either, I was wandering why it still got loaded into the packages. The package seems fine to me as it is now.

chewblacka commented:
> Hey, nice to see new contributors and contributions. I didn't knew of the per-name packaging either, I was wandering why it still got loaded into the packages. The package seems fine to me as it is now.

Much appreciated!

peterhoeg commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,66 @@
+  nativeBuildInputs = [ copyDesktopItems ];
+
+  propagatedBuildInputs = [ onedrive ] ++
+  builtins.attrValues { inherit (python3Packages) pyside6 requests; };
```
I suggest the more canonical way to write this:

```nix
  propagatedBuildInputs = [ onedrive ] ++ (with python3Packages; [ pyside6 requests ]);
```

peterhoeg commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,66 @@
+  '';
+
+  postInstall = ''
+    ln -s $src/src/{resources,ui} $out/bin/
```
This looks very strange - why are you pulling in the source?

chewblacka replied:
> This looks very strange - why are you pulling in the source?

The binary basically needs access to the subdirectories `resources` and `ui` which contain various assets. It was a choice between copying them into `$out/bin` or linking them from source. Is there a more canonical way to do this? Many thanks!

chewblacka replied:
Would this be better?

    cp -r $src/src/{resources,ui} $out/bin

peterhoeg commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,66 @@
+
+  postInstall = ''
+    ln -s $src/src/{resources,ui} $out/bin/
+    install -Dm644 src/resources/images/OneDriveGUI.png "$out/share/icons/hicolor/48x48/apps/OneDriveGUI.png"
```
2 things:

1. while 644 will work, the store is not writable, so you might as well just use 444
2. `install` has a `-t dir` argument so you don't have to repeat the actual file name

In conclusion:
```sh
install -Dm444 -t $out/share/icons/hicolor/48x48/apps src/resources/images/OneDriveGUI.png
```

peterhoeg commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,66 @@
+  postInstall = ''
+    ln -s $src/src/{resources,ui} $out/bin/
+    install -Dm644 src/resources/images/OneDriveGUI.png "$out/share/icons/hicolor/48x48/apps/OneDriveGUI.png"
+    mv -v $out/bin/OneDriveGUI.py $out/bin/onedrivegui
```
If upstream uses `OneDriveGUI.py` as the executable name, I suggest you leave it as-is and then do a symlink

peterhoeg commented:
We definitely do not want anything under $out/bin that’s not an executable.

Instead, copy those to $out/share and then patch up the executable to look in the right spot. I haven’t looked at how the program looks up it’s resources and the stuff in ui, but it might be possible to just wrap the executable and fiddle with PYTHON_PATH in the wrapper.

chewblacka commented:
> Instead, copy those to $out/share and then patch up the executable to look in the right spot. I haven’t looked at how the program looks up its resources and the stuff in ui, but it might be possible to just wrap the executable and fiddle with PYTHON_PATH in the wrapper.

Many thanks for the suggestions and help. The python script hard-codes the path it uses to look for asset folders, but it does prepend the path with `DIR_PATH` which by default it sets to the location of the binary. I've patched this using `sed` to instead point to `$out/share/OneDriveGUI` where I've placed the resource folders. This seems to work. I hope this is an acceptable solution. Many thanks!

chewblacka commented:
> We definitely do not want anything under $out/bin that’s not an executable. Instead, copy those to $out/share and then patch up the executable to look in the right spot. I haven’t looked at how the program looks up its resources and the stuff in ui, but it might be possible to just wrap the executable and fiddle with PYTHON_PATH in the wrapper.

This is the line I've added which patches the script with the correct path:

    # Patch OneDriveGUI.py so DIR_PATH poitns to shared files location
    sed -i "s|\(DIR_PATH = \)os.path.*|\1'$out/share/OneDriveGUI'|" src/OneDriveGUI.py

chewblacka commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,67 @@
+
+  preBuild = ''
+    # Patch OneDriveGUI.py so DIR_PATH poitns to shared files location
+    sed -i "s|\(DIR_PATH = \)os.path.*|\1'$out/share/OneDriveGUI'|" src/OneDriveGUI.py
```
This allows the `OneDriveGUI.py` binary to find the asset folders in `$out/share/OneDriveGUI`

peterhoeg commented:
I took the liberty of making a few changes - I suggest having a look at what has been done differently here and compare this to your version.

Feel free to adapt it to your own liking. I have added some additional comments - some of which you may want to remove.

``` nix
{ lib
, python3Packages
, fetchFromGitHub
, writeText
, copyDesktopItems
, makeDesktopItem
, makeWrapper
, onedrive
}:

let
  version = "1.0.3";

  # while we can definitely do cat $out/foo <_EOF for smaller files, it's a lot easier to deal with this outside if/when complexity increases
  setupPy = writeText "setup.py" ''
    from setuptools import setup
    setup(
      name='onedrivegui',
      version='${version}',
      scripts=[
        'src/OneDriveGUI.py',
      ],
    )
  '';

in
python3Packages.buildPythonApplication rec {
  pname = "onedrivegui";
  inherit version;

  src = fetchFromGitHub {
    owner = "bpozdena";
    repo = "OneDriveGUI";
    rev = "v${version}";
    hash = "sha256-HutziAzhIDYP8upNPieL2GNrxPBHUCVs09FFxdSqeBs=";
  };

  # makeWrapper is *technically* not needed as `buildPythonApplication` brings it in but this is
  # more explicit which makes it easier to grok (IMHO)
  nativeBuildInputs = [ copyDesktopItems makeWrapper ];

  propagatedBuildInputs = with python3Packages; [ pyside6 requests ];

  # wrap manually to avoid having a bash script in $out/bin with a .py extension
  dontWrapPythonPrograms = true;

  doCheck = false; # No tests defined

  desktopItems = [
    (makeDesktopItem {
      name = "OneDriveGUI";
      exec = "onedrivegui"; # this is the name of our wrapper
      desktopName = "OneDriveGUI";
      comment = "OneDrive GUI Client";
      type = "Application";
      icon = "OneDriveGUI";
      terminal = false;
      categories = [ "Utility" ];
    })
  ];

  postPatch = ''
    # Patch OneDriveGUI.py so DIR_PATH points to shared files location
    sed -i "s|\(DIR_PATH = \)os.path.*|\1'$out/share/OneDriveGUI'|" src/OneDriveGUI.py
    cp ${setupPy} ${setupPy.name}
  '';

  postInstall = ''
    mkdir -p $out/share/OneDriveGUI
    # we do not need the `ui` directory - only resources
    cp -r src/resources $out/share/OneDriveGUI
    install -Dm444 -t $/out/share/icons/hicolor/48x48/apps src/resources/images/OneDriveGUI.png
    # we put our own executable wrapper in place instead
    rm $out/bin/*

    makeWrapper ${python3Packages.python.interpreter} $out/bin/onedrivegui \
      --prefix PATH : ${lib.makeBinPath [ onedrive ]} \
      --prefix PYTHONPATH : ${python3Packages.makePythonPath (propagatedBuildInputs ++ [(placeholder "out")])} \
      --add-flags $out/lib/${python3Packages.python.libPrefix}/site-packages/OneDriveGUI.py
  '';

  meta = with lib; {
    homepage = "https://github.com/bpozdena/OneDriveGUI";
    description = "A simple GUI for Linux OneDrive Client, with multi-account support";
    license = licenses.gpl3Only;
    maintainers = with maintainers; [ jgarcia ];
    platforms = platforms.linux;
  };
}

```

chewblacka commented:
> I took the liberty of making a few changes - I suggest having a look at what has been done differently here and compare this to your version.
> 
> Feel free to adapt it to your own liking. I have added some additional comments - some of which you may want to remove.

Many thanks for your detailed changes! That was really good of you to go through it so thoroughly. I've kept your file mostly as is, removed one or two comments. Main change is I added an `-r` flag to `rm -r $out/bin` since `$out/bin` contains a `__pycache__` directory which errors and doesn't delete without the `-r` flag. Many thanks again. Much appreciated!

chewblacka commented:
> Please, keep the comments in the code. They may help future maintainers/users of the package.

I removed this comment:
```
# while we can definitely do cat $out/foo <_EOF for smaller files, it's a lot easier to deal with this outside if/when complexity increases
```
because it only makes sense in the context of my earlier file.

And I removed this comment:

```
  # makeWrapper is *technically* not needed as `buildPythonApplication` brings it in but this is
  # more explicit which makes it easier to grok (IMHO)
```
because it's useful to me, but I don't think is needed for future maintainers.

Janik-Haag commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,86 @@
+
+  postPatch = ''
+    # Patch OneDriveGUI.py so DIR_PATH points to shared files location
+    sed -i "s|\(DIR_PATH = \)os.path.*|\1'$out/share/OneDriveGUI'|" src/OneDriveGUI.py
```
you should use `substituteInPlace` here instead of sed

chewblacka replied:
Many thanks for those helpful suggestions! I have implemented all of them and re-pushed a commit.

chewblacka commented:
All requested changes have been applied. Package can be unblocked and merged unless there is any further changes that need applying.

Janik-Haag reviewed (approved):
> All requested changes have been applied. Package can be unblocked and merged unless there are any further changes that need applying.

Yes I know and the changes LGTM but I would wait a few days if @peterhoeg still has something

peterhoeg commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,89 @@
+python3Packages.buildPythonApplication rec {
+  pname = "onedrivegui";
+  inherit version;
+  disabled = python3Packages.pythonOlder "3.0";
```
Do we know which exact version (3.x) is needed?

chewblacka replied:
> Do we know which exact version (3.x) is needed?

No, there is no mention of minimum python version that I could find. I don't know what a sensible default would be so I just went with 3.0. Happy to set to something else.

peterhoeg commented on pkgs/by-name/on/onedrivegui/package.nix:
```diff
@@ -0,0 +1,89 @@
+
+  postPatch = ''
+    # Patch OneDriveGUI.py so DIR_PATH points to shared files location
+    substituteInPlace src/OneDriveGUI.py \
```
So @Janik-Haag made a good point about how you *in general* want to use `substituteInPlace` (and friends) rather than invoke sed directly, but in this case you run the risk of potentially substituting something unintended as this call is far less specific than what you did with sed.

I would personally just use something like this (untested):
```
sed -i src/OneDriveGUI.py \
  -e "s@^DIR_PATH =@DIR_PATH = '$out/share/OneDriveGUI@'"

peterhoeg commented:
I'm currently traveling, so a little slow to reply.

peterhoeg commented:
> No, there is no mention of minimum python version that I could find. I don’t know what a sensible default would be so I just went with 3.0. Happy to set to something else.

In that case let’s just drop it - we are building it only with py3 anyway.

There’s been a lot of churn on this PR - thank you for your work and patience!

Janik-Haag commented:
I'm sorry for the overhead I caused, should have been more careful.

peterhoeg commented:
Thanks all!
